<a href="https://colab.research.google.com/github/saikrishna9917/AIML/blob/master/assignment(11_06_20).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# CSE-221710307022-Sai Krishna

Question 1:
    https://www.codechef.com/problems/RECNDSTR



In [0]:
def L(x):
    return x[1:]+x[0]

def R(x):///
    return x[-1]+x[:len(x)-1]

n=int(input())
s=[input() for i in range(n)]
for x in s:
    print("YES") if L(x)==R(x) else print("NO")

4
a
ab
abcd
aaaaa
YES
YES
NO
YES


Question 2:
<br><a href="https://www.hackerrank.com/challenges/beautiful-days-at-the-movies/problem">Question in hackerrank</a>

In [0]:
def rev(n):
    r=0
    while(n>0):
        r=r*10+n%10
        n=n//10
    return r
s=input().split()
cou=0
s=[int(i) for i in s]
for i in range(s[0],s[1]):
    if (abs(i-rev(i)))%s[2]==0:
        cou+=1

print(cou)

20 23 6
2


Question 3: <br><a href="https://www.hackerrank.com/challenges/birthday-cake-candles/problem">Question in hackerrank</a>

In [0]:
n=int(input())
ca=input().split()
ca=[int(ca[i]) for i in range(n)]
ma=max(ca)
print(ca.count(ma))

4
3 1 2 3
2


Question 4:(InfyTQ)
Consider a non-empty array inarr containing non-zero positive single digit integers separated by “,” (comma).Print a string based on the below logic:
From all possible unique three digit number combinations with the elements of inarr
Identify the maximum number maxnum among the three digit combinations formed.
Print the maxnum followed by the total number of unique digit combinations formed,separating them with a “,” (comma)

In [0]:
import math as m
n=input().split(",")
n.sort()
a=''.join(n[::-1])
n=[int(i) for i in n]
his=dict()
for i in n:
    his[i]=his.get(i,0)+1
p=m.factorial(len(n))
k=1
for i in his.values():
    k*=m.factorial(i)
print(a+","+str(p//k))

4,1,2,1
4211,12


Question 5:(InfyTQ)
For a given positive number num,identify the palindrome formed by performing the following operations
Add num and its reverse
Check whether the sum is palindrome or not.If not,add the sum and its reverse and repeat the process until a palindrome is obtained.


In [0]:
def rev(n):
    r=0
    while(n>0):
        r=r*10+n%10
        n=n//10
    return r

n=int(input())
while True:
    r=rev(n)
    if n==r:
        print(n)
        break
    else:
        n=n+r

195
9339


Question 6:<br><a href="https://www.codechef.com/problems/SLAEL">Question in Codechef</a>

In [0]:
for i in range(int(input())):
    n,k = map(int,input().split())
    a = list(map(int,input().split()))
    c,x,i1,i2,b = 0,0,0,0,list()
    for j in range(n):
        c += 1
        if a[j] > k :
            x += 1
            if x == 1 :
                i1 = j
            else:
                if a[j] == a[i1]:
                    x -= 1
                else:i2 = j
        if x == 2 :
            x = 1
            b.append(c-1)
            c = i2-i1
            i1 = i2
            i2 = 0
        if j == n-1 and b == [ ]:b.append(n)
    print(max(b))

2
5 3
2 4 2 4 2
5
8 5
9 3 5 7 8 11 17 2
3


Question 7:
 
<br>Input string : "xyz:750,abc:265,qwerty:195"
<br>Output string : abc


In [0]:
l1=input().split(',')
l2=[i.split(':')[1] for i in l1 ]
l3=[]
for num in l2:
    n=int(num)
    rev=num[::-1]
    while n!=int(rev):
        n=n+int(rev)
        rev=str(n)[::-1]
    l3.append(n)
print(l1[l3.index(max(l3))].split(':')[0])

xyz:750,abc:265,qwerty:195
abc


Question 8:
<br>The groups_per_user function receives a dictionary.which contains group names with the list of users.Users can belong to multiple groups.Return a dictionary with the users as keys and a list of their groups as values.

Input : {"local":["admin","userA"],"public":["admin","userB"],"administrator":["admin"]}
Output :
 {'admin': ['local', 'public', 'administrator'],
 userA': ['local'],
 'userB': ['public']}


In [0]:
import re
def rev(groups):
    usergroups = {}
    for group,users in groups.items():
        for user in users:
             usergroups[user] = usergroups.get(user,[]) + [group]

    print(usergroups)


dic=input()
key=re.findall("\"(.*?)\":\[\".*?\"\]",dic)
val=re.findall("\".*?\":(\[\".*?\"\])",dic)
d=dict()
for i in range(len(key)):
    y=re.findall("\"(.*?)\"",val[i])
    d[key[i]]=y
rev(d)

{"local":["admin","userA"],"public":["admin","userB"],"administrator":["admin"]}
{'admin': ['local', 'public', 'administrator'], 'userA': ['local'], 'userB': ['public']}


Question 9:
<br>Input : london 40000 12,chicago 60000 20,munich 50000 15,paris 70000 5
<br>Output : paris munich

City                population      no.of hospitals

London	              40000	     	12

Chicago 	          60000 		20

Munich                50000 		15

Paris 		          70000 		5

In [0]:
s=input()
s1=s.split(',')
print(s1)
a={}
b={}
al=[]
bl=[]
for i in s1:
    a1=list(i.split())
    a[a1[0]]=int(a1[1])
    b[a1[0]]=int(a1[2])
a1 = sorted(a.items(), key=lambda x: x[1],reverse=True)
b1 = sorted(b.items(), key=lambda x: x[1])
for i in a1:
    al.append(i[0])
for i in b1:
    bl.append(i[0])
for i in range(0,len(a1)):
    if(al[i]==bl[i]):
        print(al[i],end=" ")

london 40000 12,chicago 60000 20,munich 50000 15,paris 70000 5
['london 40000 12', 'chicago 60000 20', 'munich 50000 15', 'paris 70000 5']
paris munich 

Question 10:
<br>Jug Filling Problem
<br>Problem statement: “Given two empty jugs with 3 and 5 liter capacities and a full jug with 8 liters, find a sequence of pours leaving four liters in the two largest jugs.”


In [0]:
cap = (8,5,3)
x = cap[0]
y = cap[1]
z = cap[2]
memory = {}
answer = []

def get_all_states(state):
    a = state[0]
    b = state[1]
    c = state[2]
    if(a==4 and b==4):
        answer.append(state)
        return True
    if((a,b,c) in memory):
        return False
    memory[(a,b,c)] = 1
    if(a>0):
        if(a+b<=y):
            if( get_all_states((0,a+b,c)) ):
                answer.append(state)
                return True
        else:
            if( get_all_states((a-(y-b), y, c)) ):
                answer.append(state)
                return True
        if(a+c<=z):
            if( get_all_states((0,b,a+c)) ):
                answer.append(state)
                return True
        else:
            if( get_all_states((a-(z-c), b, z)) ):
                answer.append(state)
                return True
    if(b>0):
        if(a+b<=x):
            if( get_all_states((a+b, 0, c)) ):
                answer.append(state)
                return True
        else:
            if( get_all_states((x, b-(x-a), c)) ):
                answer.append(state)
                return True
        if(b+c<=z):
            if( get_all_states((a, 0, b+c)) ):
                answer.append(state)
                return True
        else:
            if( get_all_states((a, b-(z-c), z)) ):
                answer.append(state)
                return True
    if(c>0):
        if(a+c<=x):
            if( get_all_states((a+c, b, 0)) ):
                ans.append(state)
                return True
        else:
            if( get_all_states((x, b, c-(x-a))) ):
                answer.append(state)
                return True
        if(b+c<=y):
            if( get_all_states((a, b+c, 0)) ):
                answer.append(state)
                return True
        else:
            if( get_all_states((a, y, c-(y-b))) ):
                answer.append(state)
                return True
    return False

initial_state = (8,0,0)
get_all_states(initial_state)
answer.reverse()
print(answer)
print(len(answer))

[(8, 0, 0), (3, 5, 0), (0, 5, 3), (5, 0, 3), (5, 3, 0), (2, 3, 3), (2, 5, 1), (7, 0, 1), (7, 1, 0), (4, 1, 3), (4, 4, 0)]
11
